digdag-standards/src/main/java/io/digdag/standards/operator/PollingWaiter.java

```diff
@@ -0,0 +1,120 @@
+    private static final String TASK = "task";
+
+    private final Config state;
+    private final String stateKey;
+
+    private final Duration minPollInterval;
+    private final Duration maxPollInterval;
+
+    private final String waitMessage;
+    private final Object[] waitMessageParameters;
+
+    private PollingWaiter(Config state, String stateKey, Duration minPollInterval, Duration maxPollInterval, String waitMessage, Object... waitMessageParameters)
+    {
+        this.state = Objects.requireNonNull(state, "state");
+        this.stateKey = Objects.requireNonNull(stateKey, "stateKey");
+        this.minPollInterval = Objects.requireNonNull(minPollInterval, "minPollInterval");
+        this.maxPollInterval = Objects.requireNonNull(maxPollInterval, "maxPollInterval");
+        this.waitMessage = Objects.requireNonNull(waitMessage, "waitMessage");
+        this.waitMessageParameters = Objects.requireNonNull(waitMessageParameters, "waitMessageParameters");
+    }
+
+    public static PollingWaiter pollingWaiter(Config state, String stateKey)
+    {
+        return new PollingWaiter(state, stateKey, DEFAULT_MIN_INTERVAL, DEFAULT_MAX_INTERVAL, DEFAULT_ERROR_MESSAGE, DEFAULT_ERROR_MESSAGE_PARAMETERS);
+    }
+
+    public PollingWaiter withWaitMessage(String waitMessage, Object... waitMessageParameters)
+    {
+        return new PollingWaiter(state, stateKey, minPollInterval, maxPollInterval, waitMessage, waitMessageParameters);
+    }
+
+    public PollingWaiter withPollInterval(Duration minPollInterval, Duration maxPollInterval)
+    {
+        return withPollInterval(DurationInterval.of(minPollInterval, maxPollInterval));
+    }
+
+    public PollingWaiter withPollInterval(DurationInterval retryInterval)
+    {
+        return new PollingWaiter(state, stateKey, retryInterval.min(), retryInterval.max(), waitMessage, waitMessageParameters);
+    }
+
+    public PollingWaiter withMinPollInterval(Duration minPollInterval)
+    {
+        return withPollInterval(minPollInterval, maxPollInterval);
+    }
+
+    public PollingWaiter withMaxPollInterval(Duration maxPollInterval)
+    {
+        return withPollInterval(minPollInterval, maxPollInterval);
+    }
+
+    public <T> T awaitOnce(Class<T> type, Function<Config, Optional<T>> f)
+    {
+        Config pollState = state.getNestedOrSetEmpty(stateKey);
+        T result = pollState.get(RESULT, type, null);
+        if (result != null) {
+            return result;
+        }
+        result = await(f);
+        pollState.set(RESULT, result);
+        return result;
+    }
+
+    public <T> T await(Function<Config, Optional<T>> f)
+    {
+        Config pollState = state.getNestedOrSetEmpty(stateKey);
+
+        Config taskState = pollState.getNestedOrSetEmpty(TASK);
+
+        Optional<T> result;
+
+        try {
+            result = f.apply(taskState);
+        }
+        catch (Exception e) {
+            throw Throwables.propagate(e);
```

Comment:
My intent was to have the `PollingRetryExecutor` handle retries separately from the `PollingWaiter` and keep waiting and retrying logic and state separate.